mypy/checkexpr.py

```diff
@@ -1143,17 +1144,23 @@ def check_overload_call(self,
                                                  callable_name=callable_name,
                                                  object_type=object_type)
                 if not unioned_errors.is_errors():
-                    # Success! Stop early.
-                    return unioned_result
+                    # Success! But we need to see maybe normal procedure gives a narrower type.
+                    union_success = True
 
         # Step 3: If the union math fails, or if there was no union in the argument types,
         #         we fall back to checking each branch one-by-one.
```

Comment:
We should probably tweak this comment -- maybe just something simple like

    # Step 3: We try checking each branch one-by-one

?

I don't think it matters too much, as long as we remove the "if the union math fails" bit.